pycaw/tools/app_mixer/callbacks_and_manager.py

```diff
@@ -0,0 +1,112 @@
+# _________________ PYCAW _________________
+from pycaw.callbacks import AudioSessionEvents, AudioSessionNotification
+
+
+# _______ SESSION MANAGER & CALLBACK _______
+class SessionManager(AudioSessionNotification):
+    """
+    as 'session_manager' referred
+    it has nothing to do with the IAudioSessionManager2 its just what it does:
+
+    handles adding and removing sessions from list.
+    handles checking if session already exists
+    by using the session.InstanceIdentifier
+    (sometimes a session fires multiple times OnSessionCreated
+    even its already there)
+
+    The session.InstanceIdentifier is unique!
+    For every new session of an app unique!
+    For multiple sessions of an app unique!
+
+    it also handles:
+    - initiating new AppRow frames and passing there the newly created session.
+    - removing sessions from list
+    - unregister all (still active) sessions from callback
+    at 'shutdown' of this app
+
+    but mainly:
+
+    on_session_created() is fired by Windows everytime a new session registers
+
+    """
+
+    @classmethod
+    def __init__(cls, RootFrameInstance, AppRowClass):
+        cls.active = {}
+        cls.RootFrameInstance = RootFrameInstance
+        cls.AppRowClass = AppRowClass
+
+    @classmethod
+    def add_session(cls, session):
+        iid = session.InstanceIdentifier
+        if iid not in cls.active:
+            AppRowInstance = cls.AppRowClass(cls.RootFrameInstance,
+                                             session,
+                                             iid,
+                                             cls)
+
+            cls.active[iid] = AppRowInstance
+
+            # _______ ADD AUDIO SESSION CALLBACK _______
+            callback = AudioSessionCallback(AppRowInstance)
+            session.register_notification(callback)
+            # add callback to be able to listen to IAudioSessionEvents
+            # pass 'app_row' (new AppRow) to
+            # AudioSessionCallback -> AudioSessionEvents,
+            # so that the callback instance can call methods from this AppRow.
+
+    @classmethod
+    def remove_session(cls, iid):
+        if iid in cls.active:
+            del cls.active[iid]
```

Comment:
We could make this one-liner with:
```suggestion
        cls.active.pop(iid, None)
```

Reply:
hold on - genius move 😇